File: ed_receiver/web/handover.mjs
import { esc, hhmm } from './view.mjs';

/** The first acknowledgement this screen recorded at or after the final packet arrived (one clock: this screen's). */
export function receivedAfterHandover(incident) {
  const handover = incident.handover;
  if (!handover) return null;
  return (incident.acknowledgements ?? []).find((ack) => ack.status === 'received' && ack.at >= handover.arrived_at) ?? null;
}

/** Who told us what, from the final packet: [{who, items}] in the vehicle's reading order (people other than the crew
 *  first, then the patient monitor, then the crew). An older vehicle does not send it: nothing is shown then. */
export function informantRows(handover) {
  return (Array.isArray(handover?.informants) ? handover.informants : [])
    .filter((row) => row && typeof row.who === 'string' && row.who.trim())
    .map((row) => ({ who: row.who.charAt(0).toUpperCase() + row.who.slice(1), items: (Array.isArray(row.items) ? row.items : []).filter(Boolean) }));
}

/** The vehicle's final packet: the handoff report frozen when the medic handed the patient over. */
export function renderHandover(incident) {
  const handover = incident.handover;
  if (!handover) return '';
  const received = receivedAfterHandover(incident);
  const sections = (handover.sections ?? []).filter((section) => (section.lines ?? []).length);
  const gaps = (handover.not_yet_known ?? []).filter(Boolean);
  const who = informantRows(handover);
  return `<section class="card handover" aria-label="Final handoff report">
    <div class="handover-head"><h2><span aria-hidden="true">✓</span> Handed over ${esc(hhmm(handover.at))} · final report</h2>
    ${received ? `<p class="handover-received" role="status"><span aria-hidden="true">✓</span> Received ${esc(hhmm(received.at))}</p>`
      : '<button id="ack-handover" class="handover-ack">Received</button>'}</div>
    ${handover.title ? `<p class="muted">${esc(handover.title)}</p>` : ''}
    ${sections.map((section) => `<section><h3>${esc(section.label)}</h3><ul>${section.lines.map((line) => `<li>${esc(line)}</li>`).join('')}</ul></section>`).join('')}
    ${who.length ? `<section class="handover-who" aria-labelledby="handover-who-h"><h3 id="handover-who-h">Who told us</h3><ul>${who.map((row) =>
      `<li><b>${esc(row.who)}</b><span>${esc(row.items.join(', ') || 'no items listed')}</span></li>`).join('')}</ul></section>` : ''}
    ${gaps.length ? `<p class="muted">Not yet known: ${esc(gaps.join(' · '))}</p>` : ''}</section>`;
}

File: herald/relay/handover.py
"""The final "handed over" packet and its delivery status.

At the hospital the medic taps Hand over once; the incident then carries `handed_over_at` and the handoff report
frozen at that moment (`handoff_final`). The relay sends that frozen report to the ED as one last packet per patient,
through the same sequenced, acknowledged, retried path as every other packet.

What leaves the vehicle is the report's section labels and line texts (built from confirmed facts only) and the
listed gaps. The report's `not_yet_confirmed` list and its read-aloud `text` (which names unconfirmed items) stay on
the vehicle: only confirmed facts reach the ED (AGENTS.md invariant 4).
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional


def _iso(value) -> Optional[str]:
    if value is None:
        return None
    return value.isoformat() if isinstance(value, datetime) else str(value)


def _parse(value) -> Optional[datetime]:
    try:
        return datetime.fromisoformat(str(value))
    except (TypeError, ValueError):
        return None


def is_handed_over(inc) -> bool:
    return bool(getattr(inc, "handed_over_at", None) and getattr(inc, "handoff_final", None))


def final_report_body(inc) -> dict:
    """The wire form of the frozen report (`ho` in the packet)."""
    report = inc.handoff_final
    fmt = report.get("format", {})
    return {
        "at": _iso(inc.handed_over_at),
        "format": fmt.get("id"), "title": fmt.get("title") or fmt.get("label"),
        "sections": [{"label": section.get("label"), "lines": [line.get("text") for line in section.get("lines", [])]}
                     for section in report.get("sections", [])],
        "not_yet_known": [gap.get("text") or gap.get("label") for gap in report.get("not_yet_known", [])],
    }


def handover_status(inc, delivered: Optional[dict], acknowledgements: list[dict]) -> Optional[dict]:
    """{"at", "delivered_at", "received_at"} for a handed-over patient, else None.

    `delivered_at`: when the ED acknowledged the final packet. `received_at`: the first clinician acknowledgement
    the ED recorded at or after the hand over."""
    at = getattr(inc, "handed_over_at", None)
    if at is None:
        return None
    received = None
    for row in acknowledgements or []:
        when = _parse(row.get("at"))
        if when is not None and when.tzinfo is not None and when >= at:
            received = row.get("at")
            break
    return {"at": _iso(at), "delivered_at": (delivered or {}).get("delivered_at"), "received_at": received}
